ESP32-S3 rendering issue

I'm experiencing issues with rendering quality and colors on the [ESP32-S3-Touch-LCD-4.3](https://www.waveshare.com/wiki/ESP32-S3-Touch-LCD-4.3)

1. the quality is very poor (with LVGL I don't have such issue)
2. the colors are completely wrong (with LVGL I don't have such issue, I also tried to invert colors through `esp_lcd_panel_invert_color` but it wasn't the issue)

I would like to point out that I'm a newbie with `Slint` so it might be my fault 

![Rendering issue](https://github.com/slint-ui/slint/assets/75322363/229aa249-556a-4ca6-a29d-076aa9aded73)

Dependency versions
```
[1/5] espressif/esp_lcd_ra8875 (1.0.0)
[2/5] espressif/esp_lcd_touch (1.1.1)
[3/5] espressif/esp_lcd_touch_gt911 (1.1.0)
[4/5] idf (5.2.0)
[5/5] slint/slint (1.6.0)
```

<details>
  <summary>slint-hello-world.cpp</summary>

  ```cpp
  #include <esp_err.h>
  #include "esp_log.h"
  #include <slint-esp.h>
  #include "driver/i2c.h"
  
  #include <esp_lcd_ra8875.h>
  #include "esp_lcd_touch_gt911.h"
  #include "esp_lcd_panel_ops.h"
  #include "esp_lcd_panel_rgb.h"
  
  #include "appwindow.h"
  
  #define I2C_MASTER_SCL_IO 9
  #define I2C_MASTER_SDA_IO 8
  #define I2C_MASTER_NUM I2C_NUM_0
  #define I2C_MASTER_FREQ_HZ 400000
  #define I2C_MASTER_TX_BUF_DISABLE 0
  #define I2C_MASTER_RX_BUF_DISABLE 0
  #define I2C_MASTER_TIMEOUT_MS 1000
  
  #define EXAMPLE_LCD_PIXEL_CLOCK_HZ (21 * 1000 * 1000)
  #define EXAMPLE_LCD_BK_LIGHT_ON_LEVEL 1
  #define EXAMPLE_LCD_BK_LIGHT_OFF_LEVEL !1
  #define EXAMPLE_PIN_NUM_BK_LIGHT -1
  #define EXAMPLE_PIN_NUM_HSYNC 46
  #define EXAMPLE_PIN_NUM_VSYNC 3
  #define EXAMPLE_PIN_NUM_DE 5
  #define EXAMPLE_PIN_NUM_PCLK 7
  #define EXAMPLE_PIN_NUM_DATA0 14  // B3
  #define EXAMPLE_PIN_NUM_DATA1 38  // B4
  #define EXAMPLE_PIN_NUM_DATA2 18  // B5
  #define EXAMPLE_PIN_NUM_DATA3 17  // B6
  #define EXAMPLE_PIN_NUM_DATA4 10  // B7
  #define EXAMPLE_PIN_NUM_DATA5 39  // G2
  #define EXAMPLE_PIN_NUM_DATA6 0   // G3
  #define EXAMPLE_PIN_NUM_DATA7 45  // G4
  #define EXAMPLE_PIN_NUM_DATA8 48  // G5
  #define EXAMPLE_PIN_NUM_DATA9 47  // G6
  #define EXAMPLE_PIN_NUM_DATA10 21 // G7
  #define EXAMPLE_PIN_NUM_DATA11 1  // R3
  #define EXAMPLE_PIN_NUM_DATA12 2  // R4
  #define EXAMPLE_PIN_NUM_DATA13 42 // R5
  #define EXAMPLE_PIN_NUM_DATA14 41 // R6
  #define EXAMPLE_PIN_NUM_DATA15 40 // R7
  #define EXAMPLE_PIN_NUM_DISP_EN -1
  
  #define EXAMPLE_LCD_H_RES 800
  #define EXAMPLE_LCD_V_RES 480
  
  #define EXAMPLE_LCD_NUM_FB 1
  
  static esp_err_t i2c_master_init()
  {
      i2c_port_t i2c_master_port = I2C_MASTER_NUM;
  
  #pragma GCC diagnostic push
  #pragma GCC diagnostic ignored "-Wmissing-field-initializers"
      i2c_config_t conf = {
          .mode = I2C_MODE_MASTER,
          .sda_io_num = I2C_MASTER_SDA_IO,
          .scl_io_num = I2C_MASTER_SCL_IO,
          .sda_pullup_en = GPIO_PULLUP_ENABLE,
          .scl_pullup_en = GPIO_PULLUP_ENABLE,
          .master = {
              .clk_speed = I2C_MASTER_FREQ_HZ},
      };
  #pragma GCC diagnostic pop
  
      i2c_param_config(i2c_master_port, &conf);
  
      return i2c_driver_install(i2c_master_port, conf.mode, I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE, 0);
  }
  
  extern "C" void app_main(void)
  {
      ESP_ERROR_CHECK(i2c_master_init());
  
      esp_lcd_touch_handle_t tp = NULL;
      esp_lcd_panel_io_handle_t tp_io_handle = NULL;
      esp_lcd_panel_io_i2c_config_t tp_io_config = ESP_LCD_TOUCH_IO_I2C_GT911_CONFIG();
      ESP_LOGI(__func__, "Initialize touch IO (I2C)");
  
      // ESP_ERROR_CHECK(esp_lcd_new_panel_io_i2c((esp_lcd_i2c_bus_handle_t)I2C_MASTER_NUM, &tp_io_config, &tp_io_handle));
      ESP_ERROR_CHECK(esp_lcd_new_panel_io_i2c_v1((esp_lcd_i2c_bus_handle_t)I2C_MASTER_NUM, &tp_io_config, &tp_io_handle));
  
  #pragma GCC diagnostic push
  #pragma GCC diagnostic ignored "-Wmissing-field-initializers"
      esp_lcd_touch_config_t tp_cfg = {
          .x_max = EXAMPLE_LCD_V_RES,
          .y_max = EXAMPLE_LCD_H_RES,
          .rst_gpio_num = (gpio_num_t)-1,
          .int_gpio_num = (gpio_num_t)4,
          .flags = {
              .swap_xy = 0,
              .mirror_x = 0,
              .mirror_y = 0,
          },
      };
  #pragma GCC diagnostic pop
  
      ESP_LOGI(__func__, "Initialize touch controller GT911");
      ESP_ERROR_CHECK(esp_lcd_touch_new_i2c_gt911(tp_io_handle, &tp_cfg, &tp));
  
      int ret;
      uint8_t write_buf = 0x01;
  
      ret = i2c_master_write_to_device(I2C_MASTER_NUM, 0x24, &write_buf, 1, I2C_MASTER_TIMEOUT_MS / portTICK_PERIOD_MS);
      ESP_LOGI(__func__, "0x48 0x01 ret is %d", ret);
  
      write_buf = 0x0E;
      ret = i2c_master_write_to_device(I2C_MASTER_NUM, 0x38, &write_buf, 1, I2C_MASTER_TIMEOUT_MS / portTICK_PERIOD_MS);
      ESP_LOGI(__func__, "0x70 0x00 ret is %d", ret);
  
      ESP_LOGI(__func__, "Install RGB LCD panel driver");
      esp_lcd_panel_handle_t panel_handle = NULL;
  
  #pragma GCC diagnostic push
  #pragma GCC diagnostic ignored "-Wmissing-field-initializers"
      esp_lcd_rgb_panel_config_t panel_config = {
          .clk_src = LCD_CLK_SRC_DEFAULT,
          .timings = {
              .pclk_hz = EXAMPLE_LCD_PIXEL_CLOCK_HZ,
              .h_res = EXAMPLE_LCD_H_RES,
              .v_res = EXAMPLE_LCD_V_RES,
              .hsync_pulse_width = 30,
              .hsync_back_porch = 30,
              .hsync_front_porch = 210,
              .vsync_pulse_width = 4,
              .vsync_back_porch = 4,
              .vsync_front_porch = 4,
              .flags = {
                  .pclk_active_neg = true,
              },
          },
          .data_width = 16,
          .num_fbs = EXAMPLE_LCD_NUM_FB,
          .bounce_buffer_size_px = 10 * EXAMPLE_LCD_H_RES,
          .psram_trans_align = 64,
          .hsync_gpio_num = EXAMPLE_PIN_NUM_HSYNC,
          .vsync_gpio_num = EXAMPLE_PIN_NUM_VSYNC,
          .de_gpio_num = EXAMPLE_PIN_NUM_DE,
          .pclk_gpio_num = EXAMPLE_PIN_NUM_PCLK,
          .disp_gpio_num = EXAMPLE_PIN_NUM_DISP_EN,
          .data_gpio_nums = {
              EXAMPLE_PIN_NUM_DATA0,
              EXAMPLE_PIN_NUM_DATA1,
              EXAMPLE_PIN_NUM_DATA2,
              EXAMPLE_PIN_NUM_DATA3,
              EXAMPLE_PIN_NUM_DATA4,
              EXAMPLE_PIN_NUM_DATA5,
              EXAMPLE_PIN_NUM_DATA6,
              EXAMPLE_PIN_NUM_DATA7,
              EXAMPLE_PIN_NUM_DATA8,
              EXAMPLE_PIN_NUM_DATA9,
              EXAMPLE_PIN_NUM_DATA10,
              EXAMPLE_PIN_NUM_DATA11,
              EXAMPLE_PIN_NUM_DATA12,
              EXAMPLE_PIN_NUM_DATA13,
              EXAMPLE_PIN_NUM_DATA14,
              EXAMPLE_PIN_NUM_DATA15,
          },
          .flags = {
              .fb_in_psram = true,
          },
      };
  #pragma GCC diagnostic pop
  
      ESP_ERROR_CHECK(esp_lcd_new_rgb_panel(&panel_config, &panel_handle));
  
      ESP_LOGI(__func__, "Initialize RGB LCD panel");
      ESP_ERROR_CHECK(esp_lcd_panel_reset(panel_handle));
      ESP_ERROR_CHECK(esp_lcd_panel_init(panel_handle));
  
      static std::vector<slint::platform::Rgb565Pixel> buffer(EXAMPLE_LCD_H_RES * EXAMPLE_LCD_V_RES);
  
      slint_esp_init(slint::PhysicalSize({EXAMPLE_LCD_H_RES, EXAMPLE_LCD_V_RES}), panel_handle,
                     tp, buffer);
  
      auto ui = AppWindow::create();
      ui->run();
  }
  ```
</details>

<details>
  <summary>appwindow.slint</summary>

```slint
import { VerticalBox, AboutSlint } from "std-widgets.slint";
export component AppWindow {
    VerticalBox {
        AboutSlint {}
        Text {
            text: "Hello World";
            font-size: 18px;
            horizontal-alignment: center;
        }
    }
}
```
</details>

<details>
  <summary>idf_component.yml</summary>

```yml
## IDF Component Manager Manifest File
dependencies:
espressif/esp_lcd_ra8875: "==1.0.0"
espressif/esp_lcd_touch_gt911: "==1.1.0"
slint/slint: "==1.6.0"
## Required IDF version
idf:
    version: ">=4.1.0"
# # Put list of dependencies here
# # For components maintained by Espressif:
# component: "~1.0.0"
# # For 3rd party components:
# username/component: ">=1.0.0,<2.0.0"
# username2/component2:
#   version: "~1.0.0"
#   # For transient dependencies `public` flag can be set.
#   # `public` flag doesn't have an effect dependencies of the `main` component.
#   # All dependencies of `main` are public by default.
#   public: true
```
</details>

That's what I'm getting in the terminal when I try to compile:

idf_component.yml
```yml
## IDF Component Manager Manifest File
dependencies:
  espressif/esp_lcd_ra8875: "==1.0.0"
  espressif/esp_lcd_touch_gt911: "==1.1.0"
  # slint/slint: "==1.6.0"
  slint:
    git: https://github.com/slint-ui/slint
    path: api/cpp/esp-idf/slint
    version: simon/esp-idf-color-swap
  ## Required IDF version
  idf:
    version: ">=4.1.0"
  # # Put list of dependencies here
  # # For components maintained by Espressif:
  # component: "~1.0.0"
  # # For 3rd party components:
  # username/component: ">=1.0.0,<2.0.0"
  # username2/component2:
  #   version: "~1.0.0"
  #   # For transient dependencies `public` flag can be set.
  #   # `public` flag doesn't have an effect dependencies of the `main` component.
  #   # All dependencies of `main` are public by default.
  #   public: true
```

terminal output
```log
 *  Executing task: cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B /Users/massimilianomontagni/Progetti/Playgrounds/esp/slint-hello-world/build -S /Users/massimilianomontagni/Progetti/Playgrounds/esp/slint-hello-world -DCCACHE_ENABLE=1 

-- Found Git: /usr/bin/git (found version "2.39.3 (Apple Git-146)") 
-- ccache will be used for faster recompilation
-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- The ASM compiler identification is GNU
-- Found assembler: /Users/massimilianomontagni/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/xtensa-esp32s3-elf-gcc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Users/massimilianomontagni/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/xtensa-esp32s3-elf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Users/massimilianomontagni/.espressif/tools/xtensa-esp-elf/esp-13.2.0_20230928/xtensa-esp-elf/bin/xtensa-esp32s3-elf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- git rev-parse returned 'fatal: not a git repository (or any of the parent directories): .git'
-- Could not use 'git describe' to determine PROJECT_VER.
-- Building ESP-IDF components for target esp32s3
Manifest files have changed, solving dependencies.
......CMake Error at /Users/massimilianomontagni/esp/v5.2/esp-idf/tools/cmake/build.cmake:544 (message):
  Traceback (most recent call last):

    File "<frozen runpy>", line 198, in _run_module_as_main
    File "<frozen runpy>", line 88, in _run_code
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/prepare_components/__main__.py", line 6, in <module>
      main()
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/prepare_components/prepare.py", line 132, in main
      args.func(args)
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/prepare_components/prepare.py", line 31, in prepare_dep_dirs
      ).prepare_dep_dirs(
        ^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/core.py", line 111, in wrapper
      return func(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/core.py", line 746, in prepare_dep_dirs
      downloaded_components = download_project_dependencies(
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/dependencies.py", line 271, in download_project_dependencies
      solution = solver.solve()
                 ^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/version_solver/version_solver.py", line 91, in solve
      self.solve_manifest(manifest)
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/version_solver/version_solver.py", line 141, in solve_manifest
      self.solve_component(dep, manifest_path=manifest.path)
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/version_solver/version_solver.py", line 161, in solve_component
      cmp_with_versions, source = self.get_versions_from_sources(requirement)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_manager/version_solver/version_solver.py", line 118, in get_versions_from_sources
      cmp_with_versions = source.versions(
                          ^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_tools/sources/git.py", line 197, in versions
      component_hash = hash_dir(source_path, include=include, exclude=exclude)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_tools/hash_tools/calculate.py", line 61, in hash_dir
      sha.update(hash_file(file_path).encode('utf-8'))
                 ^^^^^^^^^^^^^^^^^^^^
    File "/Users/massimilianomontagni/.espressif/python_env/idf5.2_py3.11_env/lib/python3.11/site-packages/idf_component_tools/hash_tools/calculate.py", line 30, in hash_file
      with open(Path(file_path).as_posix(), 'rb') as f:
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  FileNotFoundError: [Errno 2] No such file or directory:
  '/var/folders/w_/b1bc_2c96bz9dd2lck5p_jk40000gn/T/tmpxq8oos6h/api/cpp/esp-idf/slint/LICENSE'


Call Stack (most recent call first):
  /Users/massimilianomontagni/esp/v5.2/esp-idf/tools/cmake/project.cmake:605 (idf_build_process)
  CMakeLists.txt:10 (project)


-- Configuring incomplete, errors occurred!

 *  The terminal process "/bin/zsh '-l', '-c', 'cmake -G Ninja -DPYTHON_DEPS_CHECKED=1 -DESP_PLATFORM=1 -B /Users/massimilianomontagni/Progetti/Playgrounds/esp/slint-hello-world/build -S /Users/massimilianomontagni/Progetti/Playgrounds/esp/slint-hello-world -DCCACHE_ENABLE=1'" terminated with exit code: 1. 
```

--- a/api/cpp/esp-idf/slint/README.md
+++ b/api/cpp/esp-idf/slint/README.md
@@ -84,8 +84,13 @@ extern "C" void app_main(void)
     static std::vector<slint::platform::Rgb565Pixel> buffer(BSP_LCD_H_RES * BSP_LCD_V_RES);
 
     /* Initialize Slint's ESP platform support*/
-    slint_esp_init(slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }), panel_handle,
-                                       touch_handle, buffer);
+    slint_esp_init(SlintPlatformConfiguration {
+            .size = slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }),
+            .panel = panel_handle,
+            .touch = touch_handle,
+            .buffer1 = buffer,
+            .color_swap_16 = true });
+
     /* Instantiate the UI */
     auto ui = AppWindow::create();
     /* Show it on the screen and run the event loop */
@@ -167,6 +172,13 @@ using [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/ws
 
 One reason could be that you don't have enough ram for the heap or the stack.
 Make sure that the stack is big enough (~8KiB), and that all the RAM was made available for the heap allocator.
+
+### Wrong colors shown
+
+If colors look inverted on your display, it may be an incompatibility between how RGB565 colors are ordered in little-endian
+and your display expecting a different byte order. Typically, esp32 devices are little ending and display controllers often
+expect big-endian or `esp_lcd` configures them accordingly. Therefore, by default Slint converts pixels to big-endian.
+If your display controller expects little endian, set the `color_swap_16` field in `SlintPlatformConfiguration` to `false`.
 
 ## License
 

--- a/api/cpp/esp-idf/slint/include/slint-esp.h
+++ b/api/cpp/esp-idf/slint/include/slint-esp.h
@@ -95,4 +95,4 @@ void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
  *
  * This must be called before any other call to the Slint library.
  */
-void slint_esp_init(SlintPlatformConfiguration config);
+void slint_esp_init(const SlintPlatformConfiguration &config);

--- a/api/cpp/esp-idf/slint/include/slint-esp.h
+++ b/api/cpp/esp-idf/slint/include/slint-esp.h
@@ -8,9 +8,57 @@
 #include "esp_lcd_types.h"
 
 /**
+ * This data structure configures the Slint platform for use with ESP-IDF, in particular
+ * the esp_lcd component (
+ * https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/peripherals/lcd.html )
+ * for touch input and on-screen rendering.
+ *
+ * Slint supports three different ways of rendering:
+ *
+ * * Single-buffering: Allocate one frame-buffer at a location of your choosing in RAM, and
+ *                     set the `buffer1` field.
+ * * Double-buffering: Call `esp_lcd_rgb_panel_get_frame_buffer` to obtain two frame buffers
+ *                     allocated by the `esp_lcd` driver and set `buffer1` and `buffer2`.
+ * * Line-by-line rendering: Set neither `buffer1` nor `buffer2` to instruct Slint to allocate
+ *                           a buffer (with MALLOC_CAP_INTERNAL) big enough to hold one line,
+ *                           render into it, and send it to the display. This is an experimental
+ *                           feature that requires enabling the `SLINT_FEATURE_EXPERIMENTAL` CMake
+ *                           option.
+ *
+ *  Use single-buffering if you can allocate a buffer in a memory region that allows the esp_lcd
+ *  driver to efficiently transfer to the display. Use double-buffering if your driver supports
+ *  calling `esp_lcd_rgb_panel_get_frame_buffer` and the buffers can be accessed directly by the
+ *  display controller. Use line-by-line rendering if you don't have sufficient memory or rendering
+ *  to internal memory (MALLOC_CAP_INTERNAL) and flushing to the display is faster than rendering
+ *  into memory buffers that may be slower to access for the CPU.
+ */
+struct SlintPlatformConfiguration
+{
+    /// The size of the screen in pixels.
+    slint::PhysicalSize size;
+    /// The handle to the display as previously initialized by `bsp_display_new` or
+    /// `esp_lcd_panel_init`.
+    esp_lcd_panel_handle_t panel;
+    /// The touch screen handle, if the device is equipped with a touch screen.
+    std::optional<esp_lcd_touch_handle_t> touch;
+    /// The buffer Slint will render into. It must have have the size of at least one frame. Slint
+    /// calls esp_lcd_panel_draw_bitmap to flush the buffer to the screen.
+    std::optional<std::span<slint::platform::Rgb565Pixel>> buffer1 = {};
+    /// If specified, this is a second buffer that will be used for double-buffering. Use this if
+    /// your LCD panel supports double buffering: Call `esp_lcd_rgb_panel_get_frame_buffer` to
+    /// obtain two buffers and set `buffer` and `buffer2` in this data structure.
+    std::optional<std::span<slint::platform::Rgb565Pixel>> buffer2 = {};
+    slint::platform::SoftwareRenderer::RenderingRotation rotation =
+            slint::platform::SoftwareRenderer::RenderingRotation::NoRotation;
+    // Swap the 2 bytes of RGB 565 pixels before sending to the display. Use this
+    // if your CPU is little endian but the display expects big-endian.
+    bool color_swap_16 = false;
+};
+
+/**
  * Initialize the Slint platform for ESP-IDF
  *
- * This must be called before any other call to the slint library.
+ * This must be called before any other call to the Slint library.
  *
  * - `size` is the size of the screen
  * - `panel` is a handle to the display.
@@ -20,6 +68,9 @@
  * - `buffer2`, if specified, is a second buffer to be used with double buffering,
  *    both buffer1 and buffer2 should then be obtained with `esp_lcd_rgb_panel_get_frame_buffer`
  * - `rotation` applies a transformation while rendering in the buffer
+ *
+ * Note: For compatibility, this function overload selects RGB16 byte swapping if single-buffering
+ * is selected as rendering method.
  */
 void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
                     std::optional<esp_lcd_touch_handle_t> touch,
@@ -38,3 +89,10 @@ void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
                     std::optional<esp_lcd_touch_handle_t> touch,
                     slint::platform::SoftwareRenderer::RenderingRotation rotation = {});
 #endif
+
+/**
+ * Initialize the Slint platform for ESP-IDF.
+ *
+ * This must be called before any other call to the Slint library.
+ */
+void slint_esp_init(SlintPlatformConfiguration config);

--- a/api/cpp/esp-idf/slint/src/slint-esp.cpp
+++ b/api/cpp/esp-idf/slint/src/slint-esp.cpp
@@ -20,7 +20,7 @@ using RepaintBufferType = slint::platform::SoftwareRenderer::RepaintBufferType;
 
 struct EspPlatform : public slint::platform::Platform
 {
-    EspPlatform(SlintPlatformConfiguration config)
+    EspPlatform(const SlintPlatformConfiguration &config)
         : size(config.size),
           panel_handle(config.panel),
           touch_handle(config.touch),
@@ -356,7 +356,7 @@ void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
 }
 #endif
 
-void slint_esp_init(SlintPlatformConfiguration config)
+void slint_esp_init(const SlintPlatformConfiguration &config)
 {
     slint::platform::set_platform(std::make_unique<EspPlatform>(config));
 }

--- a/api/cpp/esp-idf/slint/src/slint-esp.cpp
+++ b/api/cpp/esp-idf/slint/src/slint-esp.cpp
@@ -20,18 +20,22 @@ using RepaintBufferType = slint::platform::SoftwareRenderer::RepaintBufferType;
 
 struct EspPlatform : public slint::platform::Platform
 {
-    EspPlatform(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
-                std::optional<esp_lcd_touch_handle_t> touch,
-                std::optional<std::span<slint::platform::Rgb565Pixel>> buffer1,
-                std::optional<std::span<slint::platform::Rgb565Pixel>> buffer2 = {},
-                slint::platform::SoftwareRenderer::RenderingRotation rotation = {})
-        : size(size),
-          panel_handle(panel),
-          touch_handle(touch),
-          buffer1(buffer1),
-          buffer2(buffer2),
-          rotation(rotation)
+    EspPlatform(SlintPlatformConfiguration config)
+        : size(config.size),
+          panel_handle(config.panel),
+          touch_handle(config.touch),
+          buffer1(config.buffer1),
+          buffer2(config.buffer2),
+          color_swap_16(config.color_swap_16),
+          rotation(config.rotation)
     {
+#if !defined(SLINT_FEATURE_EXPERIMENTAL)
+        if (!buffer1 && !buffer2) {
+            ESP_LOGI(TAG,
+                     "FATAL: Line-by-line rendering requested for use with slint-esp is "
+                     "experimental and not enabled in this build.");
+        }
+#endif
     }
 
     std::unique_ptr<slint::platform::WindowAdapter> create_window_adapter() override;
@@ -47,6 +51,7 @@ private:
     std::optional<esp_lcd_touch_handle_t> touch_handle;
     std::optional<std::span<slint::platform::Rgb565Pixel>> buffer1;
     std::optional<std::span<slint::platform::Rgb565Pixel>> buffer2;
+    bool color_swap_16;
     slint::platform::SoftwareRenderer::RenderingRotation rotation;
     class EspWindowAdapter *m_window = nullptr;
 
@@ -210,6 +215,20 @@ void EspPlatform::run_event_loop()
                 if (buffer1) {
                     auto region = m_window->m_renderer.render(buffer1.value(),
                                                               rotated ? size.height : size.width);
+
+                    if (color_swap_16) {
+                        for (auto [o, s] : region.rectangles()) {
+                            for (int y = o.y; y < o.y + s.height; y++) {
+                                for (int x = o.x; x < o.x + s.width; x++) {
+                                    // Swap endianess to big endian
+                                    auto px = reinterpret_cast<uint16_t *>(
+                                            &buffer1.value()[y * size.width + x]);
+                                    *px = (*px << 8) | (*px >> 8);
+                                }
+                            }
+                        }
+                    }
+
                     if (buffer2) {
                         auto s = region.bounding_box_size();
                         if (s.width > 0 && s.height > 0) {
@@ -229,12 +248,6 @@ void EspPlatform::run_event_loop()
                     } else {
                         for (auto [o, s] : region.rectangles()) {
                             for (int y = o.y; y < o.y + s.height; y++) {
-                                for (int x = o.x; x < o.x + s.width; x++) {
-                                    // Swap endianess to big endian
-                                    auto px = reinterpret_cast<uint16_t *>(
-                                            &buffer1.value()[y * size.width + x]);
-                                    *px = (*px << 8) | (*px >> 8);
-                                }
                                 esp_lcd_panel_draw_bitmap(panel_handle, o.x, y, o.x + s.width,
                                                           y + 1,
                                                           buffer1->data() + y * size.width + o.x);
@@ -256,6 +269,13 @@ void EspPlatform::run_event_loop()
                         std::span<slint::platform::Rgb565Pixel> view { lb.get(),
                                                                        line_end - line_start };
                         render_fn(view);
+                        if (color_swap_16) {
+                            // Swap endianess to big endian
+                            std::for_each(view.begin(), view.end(), [](auto &rgbpix) {
+                                auto px = reinterpret_cast<uint16_t *>(&rgbpix);
+                                *px = (*px << 8) | (*px >> 8);
+                            });
+                        }
                         esp_lcd_panel_draw_bitmap(panel_handle, line_start, line_y, line_end,
                                                   line_y + 1, lb.get());
                     });
@@ -305,8 +325,19 @@ void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
                     std::optional<std::span<slint::platform::Rgb565Pixel>> buffer2,
                     slint::platform::SoftwareRenderer::RenderingRotation rotation)
 {
-    slint::platform::set_platform(
-            std::make_unique<EspPlatform>(size, panel, touch, buffer1, buffer2, rotation));
+
+    SlintPlatformConfiguration config {
+        .size = size,
+        .panel = panel,
+        .touch = touch,
+        .buffer1 = buffer1,
+        .buffer2 = buffer2,
+        .rotation = rotation,
+        // For compatibility with earlier versions of Slint, we compute the value of
+        // color_swap_16 the way it was implemented in Slint (slint-esp) <= 1.6.0:
+        .color_swap_16 = buffer2.has_value()
+    };
+    slint_esp_init(config);
 }
 
 #ifdef SLINT_FEATURE_EXPERIMENTAL
@@ -314,7 +345,18 @@ void slint_esp_init(slint::PhysicalSize size, esp_lcd_panel_handle_t panel,
                     std::optional<esp_lcd_touch_handle_t> touch,
                     slint::platform::SoftwareRenderer::RenderingRotation rotation)
 {
-    slint::platform::set_platform(std::make_unique<EspPlatform>(size, panel, touch, std::nullopt,
-                                                                std::nullopt, rotation));
+    SlintPlatformConfiguration config { .size = size,
+                                        .panel = panel,
+                                        .touch = touch,
+                                        .buffer1 = std::nullopt,
+                                        .buffer2 = std::nullopt,
+                                        .rotation = rotation,
+                                        .color_swap_16 = false };
+    slint_esp_init(config);
 }
 #endif
+
+void slint_esp_init(SlintPlatformConfiguration config)
+{
+    slint::platform::set_platform(std::make_unique<EspPlatform>(config));
+}

--- a/examples/carousel/esp-idf/main.cpp
+++ b/examples/carousel/esp-idf/main.cpp
@@ -49,8 +49,12 @@ extern "C" void app_main(void)
 
     static std::vector<slint::platform::Rgb565Pixel> buffer(BSP_LCD_H_RES * BSP_LCD_V_RES);
 
-    slint_esp_init(slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }), panel_handle,
-                   touch_handle, buffer);
+    slint_esp_init(SlintPlatformConfiguration {
+            .size = slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }),
+            .panel = panel_handle,
+            .touch = touch_handle,
+            .buffer1 = buffer,
+            .color_swap_16 = true });
 
     auto carousel_demo = MainWindow::create();
 

--- a/examples/printerdemo_mcu/esp-idf/main/main.cpp
+++ b/examples/printerdemo_mcu/esp-idf/main/main.cpp
@@ -51,8 +51,12 @@ extern "C" void app_main(void)
 
     static std::vector<slint::platform::Rgb565Pixel> buffer(BSP_LCD_H_RES * BSP_LCD_V_RES);
 
-    slint_esp_init(slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }), panel_handle,
-                   touch_handle, buffer);
+    slint_esp_init(SlintPlatformConfiguration {
+            .size = slint::PhysicalSize({ BSP_LCD_H_RES, BSP_LCD_V_RES }),
+            .panel = panel_handle,
+            .touch = touch_handle,
+            .buffer1 = buffer,
+            .color_swap_16 = true });
 
     auto printer_demo = MainWindow::create();
     printer_demo->set_ink_levels(std::make_shared<InkLevelModel>());

